# OpenVaccine: COVID-19 mRNA Vaccine Degradation Prediction Getting Started Notebook

A starter notebook for those planning to build your submission from scratch. Will update with EDA as and when time permits.

In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

import json
from pandas.io.json import json_normalize

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
#for dirname, _, filenames in os.walk('/kaggle/input'):
#    for filename in filenames:
#        print(os.path.join(dirname, filename))

# You can write up to 5GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

## Read train and test json

In [4]:
train = pd.read_json('../train.json', lines=True)
train.T

0     \
SN_filter                                                            1   
deg_50C              [0.6382, 3.4773, 0.9988, 1.3228, 0.78770000000...   
deg_Mg_50C           [0.35810000000000003, 2.9683, 0.2589, 1.4552, ...   
deg_Mg_pH10          [0.7556, 2.983, 0.2526, 1.3789, 0.637600000000...   
deg_error_50C        [0.2167, 0.34750000000000003, 0.188, 0.2124, 0...   
deg_error_Mg_50C     [0.1501, 0.275, 0.0947, 0.18660000000000002, 0...   
deg_error_Mg_pH10    [0.26130000000000003, 0.38420000000000004, 0.1...   
deg_error_pH10       [0.2631, 0.28600000000000003, 0.0964, 0.1574, ...   
deg_pH10             [2.3375, 3.5060000000000002, 0.3008, 1.0108, 0...   
id                                                        id_001f94081   
index                                                                0   
predicted_loop_type  EEEEESSSSSSHHHHHHHSSSSBSSXSSIIIIISSIISSSSSSHHH...   
reactivity           [0.3297, 1.5693000000000001, 1.1227, 0.8686, 0...   
reactivity_error     [0.1359, 0.20700000000000002, 0.1633, 0.1452, ...   
seq_length                                                         107   
seq_scored                                                          68   
sequence             GGAAAAGCUCUAAUAACAGGAGACUAGGACUACGUAUUUCUAGGUA...   
signal_to_noise                                                  6.894   
structure            .....((((((.......)))).)).((.....((..((((((......   

                                                                  1     \
SN_filter                                                            0   
deg_50C              [7.6692, 0.0, 10.9561, 0.0, 0.0, 0.0, 0.0, 0.0...   
deg_Mg_50C           [4.8511, 4.0426, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,...   
deg_Mg_pH10          [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   
deg_error_50C        [15.4857, 7.9596, 13.3957, 5.8777, 5.8777, 5.8...   
deg_error_Mg_50C     [16.6174, 13.868, 8.1968, 8.1968, 8.1968, 8.19...   
deg_error_Mg_pH10    [73705.3985, 73705.3985, 73705.3985, 73705.398...   
deg_error_pH10       [10.1986, 9.2418, 5.0933, 5.0933, 5.0933, 5.09...   
deg_pH10             [4.947, 4.4523, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   
id                                                        id_0049f53ba   
index                                                                1   
predicted_loop_type  EEEEESSSSSSSSSSSSSSSSSSSSSSSHHHHSSSSSSSSSSBSSS...   
reactivity           [0.0, 0.0, 0.0, 2.2965, 0.0, 0.0, 0.0, 0.0, 0....   
reactivity_error     [2.8272, 2.8272, 2.8272, 4.7343, 2.5676, 2.567...   
seq_length                                                         107   
seq_scored                                                          68   
sequence             GGAAAAAGCGCGCGCGGUUAGCGCGCGCUUUUGCGCGCGCUGUACC...   
signal_to_noise                                                  0.193   
structure            .....(((((((((((((((((((((((....)))))))))).)))...   

                                                                  2     \
SN_filter                                                            1   
deg_50C              [0.9501000000000001, 1.7974999999999999, 1.499...   
deg_Mg_50C           [0.5163, 1.6823000000000001, 1.0426, 0.7902, 0...   
deg_Mg_pH10          [0.2504, 1.4021, 0.9804, 0.49670000000000003, ...   
deg_error_50C        [0.14980000000000002, 0.1761, 0.1517, 0.116700...   
deg_error_Mg_50C     [0.1033, 0.1464, 0.1126, 0.09620000000000001, ...   
deg_error_Mg_pH10    [0.1365, 0.2237, 0.1812, 0.1333, 0.1148, 0.160...   
deg_error_pH10       [0.17020000000000002, 0.178, 0.111, 0.091, 0.0...   
deg_pH10             [2.243, 2.9361, 1.0553, 0.721, 0.6396000000000...   
id                                                        id_006f36f57   
index                                                                2   
predicted_loop_type  EEEEESSSSISSIIIIISSSSMSSSHHHHHSSSMMSSSSHHHHHHS...   
reactivity           [0.44820000000000004, 1.4822, 1.1819, 0.743400...   
reactivity_error     [0.0931, 0.13290000000000002, 0.11280000000000.

Not so intuitive, right. Lets see if we can flatten it.

One thing to note is that for each parameter(5 nos) that is to be output by the model, two lists each of size seq_scored is provided for the parameter and the error in that parameter.

In [7]:
train_json = [json.loads(line) for line in open('../train.json', 'r')]

AttributeError: 'dict' object has no attribute 'loads'

In [6]:
for json in train_json:
    json['step'] = list(range(json['seq_scored']))
    json['sequence'] = list(json['sequence'])
    json['structure'] = list(json['structure'])
    json['predicted_loop_type'] = list(json['predicted_loop_type'])
    json['unpaired_probability'] = list(sum(np.load('../bpps/'+json['id']+'.npy')))

In [8]:
train_json[0]

{'index': 0,
 'id': 'id_001f94081',
 'sequence': ['G',
  'G',
  'A',
  'A',
  'A',
  'A',
  'G',
  'C',
  'U',
  'C',
  'U',
  'A',
  'A',
  'U',
  'A',
  'A',
  'C',
  'A',
  'G',
  'G',
  'A',
  'G',
  'A',
  'C',
  'U',
  'A',
  'G',
  'G',
  'A',
  'C',
  'U',
  'A',
  'C',
  'G',
  'U',
  'A',
  'U',
  'U',
  'U',
  'C',
  'U',
  'A',
  'G',
  'G',
  'U',
  'A',
  'A',
  'C',
  'U',
  'G',
  'G',
  'A',
  'A',
  'U',
  'A',
  'A',
  'C',
  'C',
  'C',
  'A',
  'U',
  'A',
  'C',
  'C',
  'A',
  'G',
  'C',
  'A',
  'G',
  'U',
  'U',
  'A',
  'G',
  'A',
  'G',
  'U',
  'U',
  'C',
  'G',
  'C',
  'U',
  'C',
  'U',
  'A',
  'A',
  'C',
  'A',
  'A',
  'A',
  'A',
  'G',
  'A',
  'A',
  'A',
  'C',
  'A',
  'A',
  'C',
  'A',
  'A',
  'C',
  'A',
  'A',
  'C',
  'A',
  'A',
  'C'],
 'structure': ['.',
  '.',
  '.',
  '.',
  '.',
  '(',
  '(',
  '(',
  '(',
  '(',
  '(',
  '.',
  '.',
  '.',
  '.',
  '.',
  '.',
  '.',
  ')',
  ')',
  ')',
  ')',
  '.',
  ')',
  ')',
  '.',
  '(',


In [9]:
train = pd.json_normalize(data = train_json, 
                            record_path ='reactivity',  
                            meta =['id','signal_to_noise',
                                  'SN_filter','seq_length','seq_scored']) 
train.rename(columns={0:'reactivity'}, inplace=True)
train['step'] = pd.json_normalize(data = train_json, 
                            record_path ='step'
                                        )
train['unpaired_probability'] = pd.json_normalize(data = train_json, 
                            record_path ='unpaired_probability'
                                        )
train['sequence'] = pd.json_normalize(data = train_json, 
                            record_path ='sequence'
                                        )
train['structure'] = pd.json_normalize(data = train_json, 
                            record_path ='structure'
                                        )
train['predicted_loop_type'] = pd.json_normalize(data = train_json, 
                            record_path ='predicted_loop_type'
                                        )
train['reactivity_error'] = pd.json_normalize(data = train_json, 
                            record_path ='reactivity_error'
                                        )
train['deg_Mg_pH10'] = pd.json_normalize(data = train_json, 
                            record_path ='deg_Mg_pH10'
                                        )
train['deg_error_Mg_pH10'] = pd.json_normalize(data = train_json, 
                            record_path ='deg_error_Mg_pH10'
                                        )
train['deg_pH10'] = pd.json_normalize(data = train_json, 
                            record_path ='deg_pH10',
                                        )
train['deg_error_pH10'] = pd.json_normalize(data = train_json, 
                            record_path ='deg_error_pH10',
                                        )
train['deg_Mg_50C'] = pd.json_normalize(data = train_json, 
                            record_path ='deg_Mg_50C',
                                        )
train['deg_error_Mg_50C'] = pd.json_normalize(data = train_json, 
                            record_path ='deg_error_Mg_50C',
                                        )
train['deg_50C'] = pd.json_normalize(data = train_json, 
                            record_path ='deg_50C',
                                        )
train['deg_error_50C'] = pd.json_normalize(data = train_json, 
                            record_path ='deg_error_50C',
                                        )

train.set_index(['id','step'], inplace=True)
train.T

AttributeError: module 'pandas' has no attribute 'json_normalize'

In [10]:
train.describe()

SN_filter        index  seq_length  seq_scored  signal_to_noise
count  2400.000000  2400.000000      2400.0      2400.0      2400.000000
mean      0.662083  1199.500000       107.0        68.0         4.530456
std       0.473099   692.964646         0.0         0.0         2.835142
min       0.000000     0.000000       107.0        68.0        -0.103000
25%       0.000000   599.750000       107.0        68.0         2.391000
50%       1.000000  1199.500000       107.0        68.0         4.442500
75%       1.000000  1799.250000       107.0        68.0         6.294250
max       1.000000  2399.000000       107.0        68.0        17.194000

Repeat the same with test set

In [ ]:
test = pd.read_json('../test.json', lines=True)
test.T

In [ ]:
test.describe()

### Make dummy submission File

This will be overwritten if everything works fine.

In [ ]:
submission = pd.read_csv('../sample_submission.csv', index_col= 0)
submission.to_csv('submission.csv')
submission.T

You have to make 107 to 130 predictions per input for the 5 parameters (only 3 of these are used for scoring).

## Develop Your model

Seems you will need something like an LSTM that works on sequence data.

In [ ]:
# Dummy

## Train your model

In [ ]:
# Dummy

## Perform your prediction

In [ ]:
# Dummy

## Filling dummy values

Lets use values mean and std from train data as a dummy submission

In [ ]:
train.describe()

In [ ]:
n = submission.shape[0]
reactivity_bar = 0.374922; reactivity_sd = 0.725652
deg_Mg_pH10_bar = 0.446303; deg_Mg_pH10_sd = 0.704172
deg_pH10_bar = 0.446911; deg_pH10_sd = 1.285747
deg_Mg_50C_bar = 0.407030; deg_Mg_50C_sd = 0.868013
deg_50C_bar = 0.425889; deg_50C_sd = 1.122356
submission['reactivity']=np.random.normal(reactivity_bar, reactivity_sd, size=n)
submission['deg_Mg_pH10']=np.random.normal(deg_Mg_pH10_bar, deg_Mg_pH10_sd, size=n)
submission['deg_pH10']=np.random.normal(deg_pH10_bar, deg_pH10_sd, size=n)
submission['deg_Mg_50C']=np.random.normal(deg_Mg_50C_bar, deg_Mg_50C_sd, size=n)
submission['deg_50C']=np.random.normal(deg_50C_bar, deg_50C_sd, size=n)
submission

## Submit

In [ ]:
submission.to_csv('submission.csv')

# EOF